homeassistant/components/stream/core.py

```diff
@@ -3,45 +3,54 @@
 
 import asyncio
 from collections import deque
-import io
-from typing import TYPE_CHECKING, Any, Callable
+import datetime
+from typing import Any, Callable
 
 from aiohttp import web
+import async_timeout
 import attr
 
 from homeassistant.components.http import HomeAssistantView
 from homeassistant.core import HomeAssistant, callback
 from homeassistant.helpers.event import async_call_later
 from homeassistant.util.decorator import Registry
 
-from .const import ATTR_STREAMS, DOMAIN
-
-if TYPE_CHECKING:
-    import av.container
-    import av.video
+from .const import ATTR_STREAMS, DOMAIN, MIN_SEGMENT_DURATION
 
 PROVIDERS = Registry()
 
 
-@attr.s
-class StreamBuffer:
-    """Represent a segment."""
+@attr.s(slots=True)
+class Part:
+    """Represent a segment part."""
 
-    segment: io.BytesIO = attr.ib()
-    output: av.container.OutputContainer = attr.ib()
-    vstream: av.video.VideoStream = attr.ib()
-    astream = attr.ib(default=None)  # type=Optional[av.audio.AudioStream]
+    duration: float = attr.ib()
+    independent: bool = attr.ib()
+    # http_start is offset from the first_moof_loc (not the start of the mp4 file)
+    # this is essentially in "http range" relative terms
+    http_range_start: int = attr.ib()
+    data: bytes = attr.ib()
 
 
-@attr.s
+@attr.s(slots=True)
 class Segment:
```

Comment:
The refactoring in Segment and Part is primarily to replace the `segment` BytesIO object with `Segment.init` and `Part.data` bytes.
The rest of the new attributes are metadata that allow keeping track of how the data is distributed between `Part`s and how far along in the segment the worker has processed.

Reply:
I'm not an expert in that either but I don't think that the overhead would be significant. I can make that change.

Reply:
I added methods to join the bytes to the Segment class. It makes sense for the recorder usage (although I leave it up to the recorder to create the BytesIO from the bytes since that is tied to the pyav method).
It also used it for one line in the `hls.py`, but I'm not sure if that's as useful. Every request in to hls has to look up the data in the Segment/Part, so there's a natural coupling between the data class structure and the serving methods. Have another look and let me know what you think.
As for the get_codec_string, the function is quite long and I don't see a huge benefit in moving it into the class either (I guess that's why you called it extreme).

Reply:
I actually was able to move most of stuff that touched the Segment metadata from the hls classes to Segment. See if you find it clearer.